bgpd/bgp_vpn.c

```diff
@@ -79,170 +79,133 @@ int show_adj_route_vpn(struct vty *vty, struct peer *peer,
-								"routeDistinguisher",
-								buffer);
-						} else {
-							vty_out(vty,
-								"Route Distinguisher: ");
-
-							if (type == RD_TYPE_AS
-							    || type == RD_TYPE_AS4)
-								vty_out(vty,
-									"%u:%d",
-									rd_as.as,
-									rd_as.val);
-							else if (type
-								 == RD_TYPE_IP)
-								vty_out(vty,
-									"%s:%d",
-									inet_ntoa(
-										rd_ip.ip),
-									rd_ip.val);
+				if (use_json) {
+					char buffer[BUFSIZ];
+					if (type == RD_TYPE_AS
+					    || type == RD_TYPE_AS4)
+						sprintf(buffer, "%u:%d",
+							rd_as.as, rd_as.val);
+					else if (type == RD_TYPE_IP)
+						sprintf(buffer, "%s:%d",
+							inet_ntoa(rd_ip.ip),
+							rd_ip.val);
+					json_object_string_add(
+						json_routes,
+						"routeDistinguisher", buffer);
+				} else {
+					vty_out(vty, "Route Distinguisher: ");
+
+					if (type == RD_TYPE_AS
+					    || type == RD_TYPE_AS4)
+						vty_out(vty, "%u:%d", rd_as.as,
+							rd_as.val);
+					else if (type == RD_TYPE_IP)
+						vty_out(vty, "%s:%d",
+							inet_ntoa(rd_ip.ip),
+							rd_ip.val);
 #if ENABLE_BGP_VNC
-							else if (
-								type
-								== RD_TYPE_VNC_ETH)
-								vty_out(vty,
-									"%u:%02x:%02x:%02x:%02x:%02x:%02x",
-									rd_vnc_eth
-										.local_nve_id,
-									rd_vnc_eth
-										.macaddr
-										.octet[0],
-									rd_vnc_eth
-										.macaddr
-										.octet[1],
-									rd_vnc_eth
-										.macaddr
-										.octet[2],
-									rd_vnc_eth
-										.macaddr
-										.octet[3],
-									rd_vnc_eth
-										.macaddr
-										.octet[4],
-									rd_vnc_eth
-										.macaddr
-										.octet[5]);
+					else if (type == RD_TYPE_VNC_ETH)
+						vty_out(vty,
+							"%u:%02x:%02x:%02x:%02x:%02x:%02x",
+							rd_vnc_eth.local_nve_id,
+							rd_vnc_eth.macaddr
+								.octet[0],
+							rd_vnc_eth.macaddr
+								.octet[1],
+							rd_vnc_eth.macaddr
+								.octet[2],
+							rd_vnc_eth.macaddr
+								.octet[3],
+							rd_vnc_eth.macaddr
+								.octet[4],
+							rd_vnc_eth.macaddr
+								.octet[5]);
 #endif
 
-							vty_out(vty, "\n");
-						}
-						rd_header = 0;
-					}
-					if (use_json) {
-						char buf_a[BUFSIZ];
-						char buf_b[BUFSIZ];
-
-						sprintf(buf_a, "%s/%d",
-							inet_ntop(rm->p.family,
-								  rm->p.u.val,
-								  buf_b,
-								  BUFSIZ),
-							rm->p.prefixlen);
-						json_object_object_add(
-							json_routes, buf_a,
-							json_array);
-					} else {
-						route_vty_out_tmp(
-							vty, &rm->p, attr,
-							SAFI_MPLS_VPN, use_json,
-							json_array);
-					}
+					vty_out(vty, "\n");
 				}
+				rd_header = 0;
+			}
+			if (use_json) {
+				char buf_a[BUFSIZ];
+				char buf_b[BUFSIZ];
+
+				sprintf(buf_a, "%s/%d",
+					inet_ntop(rm->p.family, rm->p.u.val,
+						  buf_b, BUFSIZ),
+					rm->p.prefixlen);
+				json_object_object_add(json_routes, buf_a,
+						       json_array);
+			} else {
+				route_vty_out_tmp(vty, &rm->p, binfo->attr,
```

Comment:
is the change from attr to binfo->attr a (very reasonable) speculative fix, or is it a confirmed fix?

Reply:
There is nothing else it could be.  There is no place that we store an attribute off of the `struct bgp_node`->info pointer.